doc/design/proposal-synchronous-hooks.md

```diff
@@ -0,0 +1,331 @@
+
+Example mocking require-in-the-middle:
+
+```js
+/**
+ * @typedef {{
+ *  format: string,
+ *  source: string | Buffer,
+ * }} ModuleExportsContext
+ */
+/**
+ * @typedef {{
+*   exports: any
+* }} ModuleExportsResult
+*/
+/**
+ * @param {string} url
+ * @param {ModuleExportsContext} context
+ * @param {(context: ModuleExportsContext) => {ModuleExportsResult}} nextExports
+ * @returns {ModuleExportsResult}
+ */
+function exports(url, context, nextExports) {
+  let exported = nextExports(url, context);
+  const stats = getStats(url);
+  if (stats.name && modules.includes(stats.name)) {
+    const newExports = userCallback(originalExports, stats.name, stats.basedir);
+    return { exports: newExports }
+  }
+  return {
+    exports: originalExports
+  };
+}
+```
+
+If the module loaded is CJS (`context.format` is `"commonjs"`), If the default step is run via `nextExports` run but the `exports` returned later is not reference equal to the original exports object, it will only affect later `module.exports` access in the original module, but does not affect contextual `exports.foo` accesses within the module. There are two options to address this:
+
+1. We allow the default `exports` step to take an optional `context.exports`. If it's configured, it will be used during the wrapper invocation, which unifies the contextual access. Hooks will need to use a proxy if they want the new exports object to be based on the result from default evaluation (since by the time the new overridden exports object is created, the module hasn't been evaluated yet).
+2. We expose the compiled CJS wrapper function in `context` so that users get to invoke it themselves and skip the default `exports` step.
+
+1 may provide better encapsulation, because with 2 would mean that changes to the CJS wrapper function parameters becomes breaking.
+
+If the module loaded is ESM (`context.format` is `"module"`) all the direct modification to `exports` are no-ops because ESM namespaces are not mutable. Returning a new `exports` for `require(esm)` only affects the caller of `require(esm)`, unless the hook returns a proxy to connect the changes between the new exports object and the module namespace object.
+
+## `link` (import-only): from source code to compiled module records
```

Comment:
Can you explain the use cases and motivation for link here more carefully?

The current justification used is import-in-the-middle, but this can be implemented just as well using `load`, and a unique URL resolver case to handle importing the original under a special key. So I'd be interested to hear why that approach is deemed insufficient here.

My concern is that if we make the experimental API of `vm.SourceTextModule` part of the public loader interface, then we can't claim the loaders API itself is stable.

Reply:
I wouldn't say `import-in-the-middle` "can be implemented just as well using `load`" when it has been having _constant_ issues with crashes due to needing to parse a moving target syntax rather than just operating on runtime values. We _definitely_ would like to have a more usable API like this to solve our patching problems. APMs continue to not officially support ESM due to the significant stability concerns.

Reply:
I also just realized that this is necessary for hooks to support arbitrary module types e.g. FFI, hooks need to create their own synthetic modules since Node.js has no idea how to convert those some bytes to a module.

Reply:
Wrapper modules are a well-defined technique at this point which can fully support the FFI and mocking use cases is my point, so it depends what definition of "necessary" we mean. Point taken from @Qard though that stability and small frictions in general are very much valid concerns though.

Reply:
Doesn't that mean hooks would have to layer their own transform on top of each other with the original source if they all want to wrap the modules somehow? That seems very easy to get wrong, and would incur repeated parsing overheads

Reply:
> Wrapper modules are a well-defined technique at this point 

The issue with wrapper modules appears to be exported primitives. Primitives are exported as references so setting them changes their value in the module where they're originally exported from. Unless I'm missing something, there's no way to wrap primitive exports while keeping them as references.

The `import-in-the-middle` code currently breaks a number of libraries (including `openai`) because the wrapping of primitive exports results in them being copied.

Reply:
Here's the `import-in-the-middle` GitHub issue that elaborates on primitive problem Tim brings up: https://github.com/nodejs/import-in-the-middle/issues/38

Reply:
I think without changes to the ESM spec to make the exports mutable somehow or to allow some kind of self-reflection in the patched module to interop with the patcher, the copy issue would be hard to avoid, and there isn't much we can do in Node.js about this. But one way to get around it on an API level for import-in-the-middle may be to provide an option for the consumers, so that they can choose what exports should be re-exported as-is (maintaining the live binding) and what exports should be wrapped (in the case of https://github.com/nodejs/import-in-the-middle/issues/38, the hook can choose to export `env` as-is, and wrap `setEnv` which calls out to the original `setEnv`, so that `env`'s live binding is maintained, even though it's read-only for the hooks, but that might be fine if the hooks don't actually need to wrap `env`).

Reply:
> But one way to get around it on an API level for import-in-the-middle may be to provide an option for the consumers, so that they can choose what exports should be re-exported as-is (maintaining the live binding) and what exports should be wrapped

Yes, I think inevitably this is the direction we'll be forced to go, I just wanted to highlight the fundamental issues that wrapping modules cannot solve. 

Part of what makes this tricky currently is that the ESM hooks are in a different thread and passing configuration is completely disconneced from where the actual hooks are configured in user code. To cover all possible hooks from user code, iitm is forced to wrap and hook everything at loading time. 

Looking at it from the perspective of OpenTelemetry, their plugins would need to expose loader config which would then need to be passed to the loader via `module.register()`.

Reply:
Apart from the reasons stated here, another use case of this was mentioned in https://github.com/nodejs/loaders/pull/205#issuecomment-2269707109 so I'll resolve this. In general having to parse the source texts again to do anything interesting seems to be a thing that people don't want to do, so it's a well-defined technique, yes, but suboptimal with correctness and performance concerns.